File: skeletons/facts-tool/tests/e2e/support/callgraph_run_rows.py
"""Child rows of one persisted `analyse call-graph` run."""
import sqlite3


def query(facts, sql, parameters=()):
    with sqlite3.connect(facts) as connection:
        return connection.execute(sql, parameters).fetchall()


def roots(facts, run_id):
    return query(facts, "SELECT s.qualified_name,r.usr FROM callgraph_run_root r "
                 "JOIN symbol s ON s.id=r.symbol_id WHERE r.run_id=? "
                 "ORDER BY s.qualified_name,r.usr", (run_id,))


def target(facts, run_id):
    rows = query(facts, "SELECT s.qualified_name,t.usr FROM callgraph_run_target "
                 "t JOIN symbol s ON s.id=t.symbol_id WHERE t.run_id=?", (run_id,))
    return rows[0] if rows else None


def edges(facts, run_id):
    """Reached edges as dicts with resolved source/target qualified names."""
    keys = ("source", "target", "kind", "position", "file_id", "offset", "depth",
            "cycle", "source_id", "target_id")
    rows = query(facts, "SELECT s.qualified_name,d.qualified_name,e.kind,"
                 "e.position,e.file_id,e.offset,e.depth,e.cycle,e.source_id,"
                 "e.destination_id FROM callgraph_run_edge e JOIN symbol s ON "
                 "s.id=e.source_id JOIN symbol d ON d.id=e.destination_id "
                 "WHERE e.run_id=? ORDER BY e.depth,s.qualified_name,"
                 "d.qualified_name,e.kind,e.position,e.file_id,e.offset",
                 (run_id,))
    return [dict(zip(keys, row)) for row in rows]


def edge_names(facts, run_id):
    return {(edge["source"], edge["target"]) for edge in edges(facts, run_id)}


def frontier(facts, run_id):
    return query(facts, "SELECT s.qualified_name,f.reason FROM "
                 "callgraph_run_frontier f JOIN symbol s ON s.id=f.symbol_id "
                 "WHERE f.run_id=? ORDER BY s.qualified_name,f.reason", (run_id,))


def recovery(facts, run_id):
    return query(facts, "SELECT tu_file_id,outcome,diagnostic FROM "
                 "callgraph_run_recovery WHERE run_id=? ORDER BY tu_file_id",
                 (run_id,))


def child_row_count(facts, run_id):
    return sum(query(facts, f"SELECT COUNT(*) FROM {table} WHERE run_id=?",
                     (run_id,))[0][0]
               for table in ("callgraph_run_root", "callgraph_run_target",
                             "callgraph_run_edge", "callgraph_run_frontier",
                             "callgraph_run_recovery"))

File: skeletons/facts-tool/src/commands/analyse/CallGraphRunRows.cpp
#include "commands/analyse/CallGraphRunRows.h"

#include "storage/catalog/Database.h"

#include <chrono>
#include <format>
#include <numeric>

namespace facts::commands {
namespace {
std::string joinComponents(const std::vector<std::string> &components) {
  return std::accumulate(components.begin(), components.end(), std::string{},
                         [](std::string joined, const std::string &name) {
                           return joined.empty() ? name : joined + "," + name;
                         });
}

std::string utcNow() {
  return std::format("{:%FT%TZ}", std::chrono::floor<std::chrono::seconds>(
                                      std::chrono::system_clock::now()));
}

template <typename Value>
std::optional<std::int64_t> integer(const std::optional<Value> &value) {
  if (!value)
    return std::nullopt;
  return static_cast<std::int64_t>(*value);
}

std::optional<std::string> text(const std::string &value) {
  return value.empty() ? std::nullopt : std::optional{value};
}
} // namespace

std::expected<std::int64_t, std::string>
insertCallGraphRun(storage::Database &database,
                   const CallGraphRunRecord &record) {
  const auto pathMode = record.pathMode.transform(
      [](auto mode) { return std::string{callgraph::pathModeName(mode)}; });
  const auto time = record.limits.time.transform(
      [](auto value) { return static_cast<std::int64_t>(value.count()); });
  return catalog::execute(
             database,
             "INSERT INTO callgraph_run(created_at,project_path,facts_path,"
             "mode,path_mode,calls_scope,components,max_depth,max_nodes,"
             "max_edges,time_limit_ms,recover_missing,status,"
             "truncation_reason,error) VALUES(?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)",
             utcNow(), record.projectPath, record.factsPath,
             callgraph::queryModeName(record.mode), pathMode,
             callgraph::scopeName(record.callsScope),
             joinComponents(record.components), integer(record.limits.depth),
             integer(record.limits.nodes), integer(record.limits.edges), time,
             record.recoverMissing ? 1 : 0, runStatusName(record.status),
             text(record.truncationReason), text(record.error))
      .transform([&] {
        return static_cast<std::int64_t>(
            sqlite3_last_insert_rowid(database.nativeHandle()));
      });
}

std::expected<void, std::string>
insertCallGraphRunRows(storage::Database &database, std::int64_t runId,
                       const CallGraphRunRecord &record) {
  for (const auto &[id, usr] : record.roots)
    if (auto row = catalog::execute(
            database, "INSERT INTO callgraph_run_root VALUES(?,?,?)", runId,
            id, usr);
        !row)
      return row;
  if (record.target)
    if (auto row = catalog::execute(
            database, "INSERT INTO callgraph_run_target VALUES(?,?,?)", runId,
            record.target->first, record.target->second);
        !row)
      return row;
  for (const auto &value : record.edges)
    if (auto row = catalog::execute(
            database,
            "INSERT OR IGNORE INTO callgraph_run_edge VALUES(?,?,?,?,?,?,?,?,?)",
            runId, value.edge.source, value.edge.destination, value.edge.kind,
            value.edge.position, value.edge.file, value.edge.offset,
            value.depth, value.cycle ? 1 : 0);
        !row)
      return row;
  for (const auto &item : record.frontier)
    if (auto row = catalog::execute(
            database, "INSERT INTO callgraph_run_frontier VALUES(?,?,?)",
            runId, item.id, item.reason);
        !row)
      return row;
  for (const auto &row : record.recovery)
    if (auto inserted = catalog::execute(
            database, "INSERT INTO callgraph_run_recovery VALUES(?,?,?,?)",
            runId, row.translationUnit, row.outcome, text(row.diagnostic));
        !inserted)
      return inserted;
  return {};
}
} // namespace facts::commands
